### PR TITLE
save headers for all intermediate redirects

### LINKED ISSUE #2656: Save headers files for all redirects during file download
There's one quirk about this method regarding redirects: we only save `headers.json` file for requested URL and one for final URL after all redirects.

Consider such scenario (with `tools/http_server.py` running):
```
download http://localhost:4548/tests/001_hello.js
redirect to http://localhost:4546/tests/001_hello.js 
redirect to http://localhost:4545/tests/001_hello.js 
fetch from http://localhost:4545/tests/001_hello.js 
SourceFile(url: "http://localhost:4545/tests/001_hello.js", redirect_source_url: "http://localhost:4548/tests/001_hello.js")
cache headers for "http://localhost:4548/tests/001_hello.js": redirect -> http://localhost:4546/tests/001_hello.js
cache file "http://localhost:4546/tests/001_hello.js"
cache headers for "http://localhost:4546/tests/001_hello.js"
```

```
download http://localhost:4546/tests/001_hello.js
redirect to http://localhost:4545/tests/001_hello.js 
fetch from http://localhost:4545/tests/001_hello.js  // !!! unnecessary fetch, we already have cached file "http://localhost:4545/tests/001_hello.js"
SourceFile(url: "http://localhost:4545/tests/001_hello.js", redirect_source_url: "http://localhost:4546/tests/001_hello.js")
```

Should we save headers for all intermediate redirects 🤔 ? 

_Originally posted by @bartlomieju in https://github.com/denoland/deno/pull/2636_

During download of remote file we should save headers file for all intermediate redirects and also after redirect it should be checked if there's headers file to handle redirect target to prevent unnecessary file download.

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #2656

### CODE REVIEW

File: cli/deno_dir.rs [hunk 1]
```diff
@@ -312,19 +312,16 @@ impl DenoDir {
     // Fetch remote file and cache on-disk for subsequent access
     // not cached/local, try remote.
     let module_url_ = module_url.clone();
-    Either::B(self
+    Either::B(
+      self
         .fetch_remote_source_async(&module_url)
         // TODO: cache fetched remote source here - `fetch_remote_source` should only fetch with
         // redirects, nothing more
```
Comment (ry) [inline cli/deno_dir.rs]: Just to confirm, this TODO still applies, right?

File: cli/deno_dir.rs [hunk 2]
```diff
@@ -421,102 +418,91 @@ impl DenoDir {
   }
 
   /// Asynchronously fetch remote source file specified by the URL following redirects.
+  // TODO: add limit of redirects param
   fn fetch_remote_source_async(
     self: &Self,
     module_url: &Url,
-  ) -> impl Future<Item = Option<SourceFile>, Error = ErrBox> {
+  ) -> Box<dyn Future<Item = SourceFile, Error = ErrBox> + Send> {
```
Comment (ry) [inline cli/deno_dir.rs]: Isn't this a `Box<SourceFileFuture>` ?

File: cli/deno_dir.rs [hunk 3]
```diff
@@ -421,102 +418,91 @@ impl DenoDir {
...
+    let dir = self.clone();
+    let module_url = module_url.clone();
+
+    // Single pass fetch, either yields code or yields redirect.
+    let f = http_util::fetch_string_once(module_uri).and_then(
+      move |fetch_once_result| {
+        match fetch_once_result {
+          FetchOnceResult::Redirect(uri) => {
+            // If redirects, update module_name and filename for next looped call.
+            let new_module_url = Url::parse(&uri.to_string()).expect("http::uri::Uri should be parseable as Url");
```
Comment (ry) [inline cli/deno_dir.rs]: This seems longer than 80 columns. Does rustfmt not wrap it?

File: cli/deno_dir.rs [hunk 4]
```diff
@@ -421,102 +418,91 @@ impl DenoDir {
...
+              None,
+              Some(new_module_url.to_string())
+            ).unwrap();
+
+            // Explicit drop to keep reference alive until future completes.
+            drop(download_job);
+
+            let module_url = module_url.clone();
+
+            // Recurse
```
Comment (ry) [inline cli/deno_dir.rs]: Is there a mechanism to avoid infinite recursion?

File: cli/deno_dir.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/deno_dir.rs]: It is, thanks :+1:

File: cli/deno_dir.rs (hunk 4 above)
Comment (bartlomieju) [reply cli/deno_dir.rs]: I added basic one just now

File: cli/deno_dir.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/deno_dir.rs]: Still fiddling with API, I'll address it before it's landable

File: cli/deno_dir.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/deno_dir.rs]: I think `tools/format.py` is not working properly after latest update... I can add indentation, run formatting and nothing happens 😢

File: cli/deno_dir.rs [hunk 5]
```diff
@@ -280,51 +280,13 @@ impl DenoDir {
...
-              std::io::ErrorKind::NotFound,
-              format!("cannot find remote file '{}'", module_url_.to_string()),
-            ).into(),
-          ),
-        }))
+    Either::B(self.fetch_remote_source_async(
+      &module_url,
+      use_disk_cache,
+      no_remote_fetch,
+      10,
```
Comment (bartlomieju) [inline cli/deno_dir.rs]: I think we should allow for 10-20 redirects CC @ry

Comment (ry) [reply cli/deno_dir.rs]: that's fine

File: cli/deno_dir.rs [hunk 6]
```diff
@@ -1064,64 +1035,127 @@ mod tests {
...
+          .unwrap();
+
+      let result =
+        deno_dir.fetch_remote_source(&double_redirect_url, false, false, 2);
+      assert!(result.is_ok());
+      let result =
+        deno_dir.fetch_remote_source(&double_redirect_url, false, false, 1);
+      assert!(result.is_err());
+
+      // TODO: add better tests
```
Comment (ry) [inline cli/deno_dir.rs]: This isn't an actionable TODO. Be more specific.

Comment (bartlomieju) [reply cli/deno_dir.rs]: Ooops, missed that one. Fixed

**Pull request merged**